prov/gni/src/gnix_sep.c

```diff
@@ -0,0 +1,860 @@
+	if ((sep_priv->ep_fid.fid.fclass != FI_CLASS_SEP) ||
+		(index >= sep_priv->info->ep_attr->tx_ctx_cnt))
+		return -FI_EINVAL;
+
+	/*
+	 * check to see if the tx context was already
+	 * allocated
+	 */
+
+	fastlock_acquire(&sep_priv->sep_lock);
+
+	if (sep_priv->tx_ep_table[index] != NULL) {
+		ret = -FI_EBUSY;
+		goto err;
+	}
+
+	tx_priv = calloc(1, sizeof(struct gnix_fid_trx));
+	if (!tx_priv) {
+		ret = -FI_ENOMEM;
+		goto err;
+	}
+
+	tx_priv->ep_fid.fid.fclass = FI_CLASS_TX_CTX;
+	tx_priv->ep_fid.fid.context = context;
+	tx_priv->ep_fid.fid.ops = &gnix_sep_fi_ops;
+	tx_priv->ep_fid.ops = &gnix_sep_ops;
+	tx_priv->ep_fid.msg = &gnix_sep_msg_ops;
+	tx_priv->ep_fid.rma = &gnix_sep_rma_ops;
+	tx_priv->ep_fid.tagged = &gnix_sep_tagged_ops;
+	tx_priv->ep_fid.atomic = &gnix_sep_atomic_ops;
+	tx_priv->ep_fid.cm = &gnix_sep_rxtx_cm_ops;
+
+	/* if an EP already allocated for this index, use it */
+	if (sep_priv->ep_table[index] != NULL) {
+		ep_priv = container_of(sep_priv->ep_table[index],
+				       struct gnix_fid_ep, ep_fid);
+		sep_priv->tx_ep_table[index] = sep_priv->ep_table[index];
+		_gnix_ref_get(ep_priv);
+	} else {
+
+		/*
+		 * allocate the underlying gnix_fid_ep struct
+		 */
+
+		ep_attr.use_cdm_id = true;
+		ep_attr.cdm_id = sep_priv->cdm_id_base + index;
+		ep_attr.cm_nic = sep_priv->cm_nic;
+		ep_attr.cm_ops = &gnix_sep_rxtx_cm_ops;
+		/* TODO: clean up this cm_nic */
+		_gnix_ref_get(sep_priv->cm_nic);
+		ret = _gnix_ep_alloc(sep_priv->domain,
+				     sep_priv->info,
+				     &ep_attr,
+				     &ep_ptr, context);
+		if (ret != FI_SUCCESS) {
+			GNIX_WARN(FI_LOG_EP_CTRL,
+				  "gnix_ep_alloc returned %s\n",
+				  fi_strerror(-ret));
+			goto err;
+		}
+
+		sep_priv->ep_table[index] = ep_ptr;
+		sep_priv->tx_ep_table[index] = ep_ptr;
+		ep_priv = container_of(ep_ptr, struct gnix_fid_ep, ep_fid);
+	}
+
+	_gnix_ref_init(&tx_priv->ref_cnt, 1, __trx_destruct);
+	tx_priv->ep = ep_priv;
+	tx_priv->sep = sep_priv;
+	_gnix_ref_get(sep_priv);
+	tx_priv->caps = ep_priv->caps;
+	*tx_ep = &tx_priv->ep_fid;
+err:
+	fastlock_release(&sep_priv->sep_lock);
+
+	return ret;
+}
+
+static int gnix_sep_rx_ctx(struct fid_ep *sep, int index,
+			   struct fi_rx_attr *attr,
+			   struct fid_ep **rx_ep, void *context)
+{
+	int ret = FI_SUCCESS;
+	struct gnix_fid_sep *sep_priv;
+	struct gnix_fid_ep *ep_priv = NULL;
+	struct gnix_fid_trx *rx_priv = NULL;
+	struct fid_ep *ep_ptr;
+	struct gnix_ep_attr ep_attr = {0};
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	sep_priv = container_of(sep, struct gnix_fid_sep, ep_fid);
+	if ((sep_priv->ep_fid.fid.fclass != FI_CLASS_SEP) ||
+		(index >= sep_priv->info->ep_attr->rx_ctx_cnt))
+		return -FI_EINVAL;
+
+	/*
+	 * check to see if the rx context was already
+	 * allocated
+	 */
+
+	fastlock_acquire(&sep_priv->sep_lock);
+
+	if (sep_priv->rx_ep_table[index] != NULL) {
+		ret = -FI_EBUSY;
+		goto err;
+	}
+
+	rx_priv = calloc(1, sizeof(struct gnix_fid_trx));
```

Comment:
Just a point of mention, but considering how often we do x = calloc(1, sizeof(y)) in the code, perhaps we should add a ZALLOC function for allocating single element, zero'd out memory allocations. .